src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/IServiceCollectionExtensions.cs

```diff
@@ -118,7 +119,9 @@ public static void AddSemanticTokensServices(this IServiceCollection services)
 
         services.AddSingleton<WorkspaceSemanticTokensRefreshPublisher, DefaultWorkspaceSemanticTokensRefreshPublisher>();
         services.AddSingleton<ProjectSnapshotChangeTrigger, DefaultWorkspaceSemanticTokensRefreshTrigger>();
-        services.AddSingleton<RazorSemanticTokensInfoService, DefaultRazorSemanticTokensInfoService>();
+
+        // Ensure that we don't add the default service if something else has added one.
+        services.TryAddSingleton<RazorSemanticTokensInfoService, DefaultRazorSemanticTokensInfoService>();
```

Comment:
This makes me wonder if we ought to be doing the `TryAddSingleton` for ALL the services. But then again maybe if we don't have a reason they ought to fail if someone adds two?

Reply:
I'd rather not... i'd rather not use DI at all in the benchmarks if we can avoid it, and just directly construct the `TestXXX` instances, but i imagine sometimes that can't be avoided.

Reply:
FWIW, the ordering changed at some point (CLaSP?). According to the history, there was a time when extra services would get added after the language server had initialized its services. Now extra services are added first, which results in the language server services always stomping on test services, at least in the micro benchmarks. I wouldn't have done this here, but then I found this:

https://github.com/dotnet/razor/blob/52a493a655a7edfae7741d0782cde8f6c0f1d03b/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageServer.cs#L151-L152

Reply:
I do agree with @davidwengier that we shouldn't use `TryAddSingleton` everywhere in the language server. I'd rather improve the infrastructure to make it easy to do the right thing.